packages/pigeon/lib/generator.dart

```diff
@@ -3,11 +3,123 @@
 // found in the LICENSE file.
 
 import 'ast.dart';
+import 'generator_tools.dart';
 
 /// A superclass of generator classes.
 ///
 /// This provides the structure that is common across generators for different languages.
 abstract class Generator<T> {
-  /// Generates files for specified language with specified [languageOptions]
-  void generate(T languageOptions, Root root, StringSink sink);
+  /// Generates files for specified language with specified [generatorOptions]
+  void generate(T generatorOptions, Root root, StringSink sink);
+}
+
+/// A Superclass that enforces code generation across platforms.
+abstract class StructuredGenerator<T> extends Generator<T> {
+  @override
+  void generate(
+    T generatorOptions,
+    Root root,
+    StringSink sink,
+  ) {
+    final Indent indent = Indent(sink);
+
+    prePrologue(generatorOptions, root, sink, indent);
+    writeFilePrologue(generatorOptions, root, sink, indent);
+    writeFileImports(generatorOptions, root, sink, indent);
+
+    preEnums(generatorOptions, root, sink, indent);
+    for (final Enum anEnum in root.enums) {
+      writeEnum(generatorOptions, root, sink, indent, anEnum);
+    }
+
+    preDataClasses(generatorOptions, root, sink, indent);
+    for (final Class klass in root.classes) {
+      writeDataClass(generatorOptions, root, sink, indent, klass);
+    }
+
+    preApis(generatorOptions, root, sink, indent);
+    for (final Api api in root.apis) {
+      if (api.location == ApiLocation.host) {
+        writeHostApi(generatorOptions, root, sink, indent, api);
+      } else if (api.location == ApiLocation.flutter) {
+        writeFlutterApi(generatorOptions, root, sink, indent, api);
+      }
+    }
+    finalWriteFile(generatorOptions, root, sink, indent);
```

Comment:
This would essentially be the same as overriding the generate method wouldn't it? The only real difference being where Indent is created.

Reply:
We should look at where we need extra bits; it might be better as something like `writeFileBody` with things like the copyright and includes and such already having been called.

Reply:
That makes sense

Reply:
I'm going to wait to see what @gaaclarke thinks about this one. I've changed it slightly to add this code to the new writeApis method for now. If adding a writeBody method is best, it's an easy change.